contracts/optimize/SlotHelper.sol

```diff
@@ -0,0 +1,159 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+// 存储方法
+library SlotHelper{
+    uint internal constant RIGHTSLOTDATASHIFT = 32;
+    uint internal constant LENOFFSET = 224;
+    uint internal constant FIRSTSLOTDATASIZE = 28;
+    
+
+    function putRaw(bytes32 key , bytes memory data)
+        internal 
+        returns(bytes32 mdata)
+    {
+        // warn: if data as ptr to move,your should keep "data.length" by another ptr
+        uint len = data.length;
+        mdata = encodeMetadata(data);
+        if (len > FIRSTSLOTDATASIZE){
+            for (uint index = 0 ; index * 32 < len - FIRSTSLOTDATASIZE ; index ++){
+                assembly{
+                    data := add(data,0x20)
+                    mstore(0,add(key,index))
+                    let slot := keccak256(0,0x20)
+
+                    //Restructured data
+                    let value1 := shl( LENOFFSET , mload(data))
+                    let value2 := shr( RIGHTSLOTDATASHIFT , mload(add(data,0x20)))
+                    value1 := or(value1,value2)
+
+                    sstore(slot,value1)
+                }
+            }
+        }
+    }
+
+    function encodeMetadata(bytes memory data) internal pure returns(bytes32 medata){
+        assembly{
+            let len := mload(data)
+            let value := mload(add(data,0x20))
+            
+            // operator in stack 
+            len := shl(LENOFFSET,len)
```

Comment:
I would prefer using non-asm code for bit operations here.  This also allows old version of EVM can use the code.

Reply:
https://github.com/QuarkChain/evm-large-storage/pull/1/commits/306e66e939cc7c1bffe0bded424d4f5bc6360fb2